src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -52,24 +54,56 @@ private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource
         {
             using IEnumerator<TSource> e = source.GetEnumerator();
 
+            // Before allocating anything, make sure there's at least one element.
             if (e.MoveNext())
             {
-                List<TSource> chunkBuilder = new();
-                int lastLength;
+                // Now that we know we have at least one item, allocate an initial storage array. This is not
+                // the array we'll yield.  It starts out small in order to avoid significantly overallocating
+                // when the source has many fewer elements than the chunk size.
+                var array = new TSource[Math.Min(size, 4)];
```

Comment:
Would it be possible to alleviate some of the throughput issues by picking a minimum that is less conservative than 4? IIRC we went for the resizing approach to address issues related to users passing chunk sizes close to or even exceeding `Array.MaxLength`.

Reply:
We can use a different value, but anything large we pick still risks significant overallocation if the input is small. What did you have in mind?

Reply:
In practice will it make a huge difference performance wise if the initial array length is 4 or 32? We might be overallocating in certain cases, but the overallocation is bounded by the requested chunk size (and a reasonable minimum anticipating pathological inputs) in a method that is otherwise expected to liberally allocate arrays.

It might be that this is a red herring, and whether we pick 4 or 32 or 1024 won't have a substantial impact on benchmarks. I haven't run any myself.

Reply:
It depends on the user-specified chunk size, which makes the hard to predict.  Whatever value we select here, it only affects the performance of the first chunk; after that, the buffer is sized appropriately and we'll take the new fast path for all subsequent chunks.  For the first chunk, though, there are two main costs:
- Whether we take the fast or slow path: we'll take the fast path only if the selected default is >= the chunk size.
- Allocation: the larger the default we select, the more reallocation associated with regrowth we potentially avoid, but the more likely it is to overallocate when the source has fewer elements.

I opted to stick with 4 as it's the tried and true value used by `List<T>` and for all source lengths > 1, ensures we never allocate more than 2x what's needed.  That said, I'm not really concerned about the performance of this method; that ship sailed the moment we settled on the `IEnumerable<T[]>` return type.  I'm just making these tweaks in this PR because I was having trouble stomach a 20% regression.  If you'd like a different default value, just name it and I'll change it accordingly.